OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs

```diff
@@ -9,44 +9,83 @@
 #endregion
 
 using System;
+using System.Collections.Generic;
 using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Default trait for rendering sprite-based actors.")]
-	class WithSpriteBodyInfo : UpgradableTraitInfo, ITraitInfo, Requires<RenderSpritesInfo>
+	public class WithSpriteBodyInfo : UpgradableTraitInfo, ITraitInfo, IRenderActorPreviewSpritesInfo, IQuantizeBodyOrientationInfo,
+		Requires<IBodyOrientationInfo>, Requires<RenderSpritesInfo>
 	{
 		[Desc("Animation to play when the actor is created.")]
 		public readonly string StartSequence = null;
 
-		[Desc("Animation to play when the actor is idle.")]
+		[Desc("Animation to play when the actor is idle/standing.")]
 		public readonly string Sequence = "idle";
 
-		public object Create(ActorInitializer init) { return new WithSpriteBody(init, this); }
+		public virtual object Create(ActorInitializer init) { return new WithSpriteBody(init, this); }
+
+		public virtual IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer init, RenderSpritesInfo rs, string image, int facings, PaletteReference p)
+		{
+			var ifacing = init.Actor.Traits.GetOrDefault<IFacingInfo>();
+			var facing = ifacing != null ? init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : ifacing.GetInitialFacing() : 0;
+
+			var anim = new Animation(init.World, image, () => facing);
+			anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence));
+
+			yield return new SpriteActorPreview(anim, WVec.Zero, 0, p, rs.Scale);
+		}
+
+		public virtual int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceProvider, string race)
+		{
+			var rsi = ai.Traits.Get<RenderSpritesInfo>();
+			return sequenceProvider.GetSequence(rsi.GetImage(ai, sequenceProvider, race), Sequence).Facings;
+		}
 	}
 
-	class WithSpriteBody : UpgradableTrait<WithSpriteBodyInfo>, ISpriteBody
+	public class WithSpriteBody : UpgradableTrait<WithSpriteBodyInfo>, ISpriteBody
 	{
-		readonly Animation body;
-		readonly WithSpriteBodyInfo info;
+		public readonly Animation DefaultAnimation;
 
-		public WithSpriteBody(ActorInitializer init, WithSpriteBodyInfo info)
+		public WithSpriteBody(ActorInitializer init, WithSpriteBodyInfo info, Func<int> baseFacing)
```

Comment:
If a subclass calls this directly then we will ignore any `StartSequence` that may be defined.